### PR TITLE
fix: Sampling: CUDA Graph fix

### PR DESCRIPTION
**Pull request by IzzyPutterman**
## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

## Summary by CodeRabbit

* **Refactor**
  * Sampling APIs now accept seed and offset as either scalars or 1D tensors; tensor-based per-call seeds/offsets are threaded through native sampling paths with scalar fallbacks preserved.
* **Quality**
  * Centralized validation/conversion enforces dtype, device, shape/length and batch-size semantics; validation applied before processing.
* **Documentation**
  * Docstrings updated with union-type signatures, usage guidance and CUDA-graph compatibility notes.

### CODE REVIEW

**Comment by yzh119** [comment]
@flashinfer-bot run

File: include/flashinfer/sampling.cuh [hunk 1]
```diff
@@ -701,8 +701,14 @@ template <uint32_t BLOCK_THREADS, BlockScanAlgorithm SCAN_ALGORITHM,
...
           typename DType, typename IdType>
 __global__ void SamplingFromLogitsKernel(DType* logits, IdType* output, IdType* indices, uint32_t d,
-                                         uint64_t philox_seed, uint64_t philox_offset) {
+                                         uint64_t* seed_arr, uint64_t seed_val,
+                                         uint64_t* offset_arr, uint64_t offset_val) {
   const uint32_t bx = blockIdx.x, tx = threadIdx.x;
+
+  // Resolve seed/offset from tensor or scalar
+  uint64_t philox_seed = seed_arr ? seed_arr[0] : seed_val;
+  uint64_t philox_offset = offset_arr ? offset_arr[0] : offset_val;
```
Comment (yzh119) [inline include/flashinfer/sampling.cuh:716]: Some block need to advance the offset value (like in https://github.com/flashinfer-ai/flashinfer/blob/6ae5bfe6a48c591a171f32888e5de28b9ca0207b/flashinfer/sampling.py#L33-L49), otherwise we will be keep using the same offset for multiple runs. Or we can advance the offset tensor after calling this operator.

File: include/flashinfer/sampling.cuh [hunk 2]
```diff
@@ -790,9 +802,15 @@ template <uint32_t BLOCK_THREADS, BlockScanAlgorithm SCAN_ALGORITHM,
...
                                            IdType* top_k_arr, uint32_t top_k_val, uint32_t d,
-                                           uint64_t philox_seed, uint64_t philox_offset) {
+                                           uint64_t* seed_arr, uint64_t seed_val,
+                                           uint64_t* offset_arr, uint64_t offset_val) {
   const uint32_t batch_size = gridDim.x;
   const uint32_t bx = blockIdx.x, tx = threadIdx.x;
+
+  // Resolve seed/offset from tensor or scalar
+  uint64_t philox_seed = seed_arr ? seed_arr[0] : seed_val;
+  uint64_t philox_offset = offset_arr ? offset_arr[0] : offset_val;
```
Comment (yzh119) [inline include/flashinfer/sampling.cuh:818]: ditto

File: include/flashinfer/sampling.cuh [hunk 3]
```diff
@@ -907,9 +925,15 @@ template <uint32_t BLOCK_THREADS, BlockScanAlgorithm SCAN_ALGORITHM,
...
                                            float* top_p_arr, float top_p_val, uint32_t d,
-                                           uint64_t philox_seed, uint64_t philox_offset) {
+                                           uint64_t* seed_arr, uint64_t seed_val,
+                                           uint64_t* offset_arr, uint64_t offset_val) {
   const uint32_t batch_size = gridDim.x;
   const uint32_t bx = blockIdx.x, tx = threadIdx.x;
+
+  // Resolve seed/offset from tensor or scalar
+  uint64_t philox_seed = seed_arr ? seed_arr[0] : seed_val;
+  uint64_t philox_offset = offset_arr ? offset_arr[0] : offset_val;
```
Comment (yzh119) [inline include/flashinfer/sampling.cuh:941]: ditto

File: include/flashinfer/sampling.cuh [hunk 4]
```diff
@@ -1018,8 +1042,14 @@ template <uint32_t BLOCK_THREADS, BlockScanAlgorithm SCAN_ALGORITHM,
...
 __global__ void MinPSamplingFromProbKernel(DType* probs, float* min_p_arr, IdType* output,
                                            IdType* indices, float min_p_val, uint32_t d,
-                                           uint64_t philox_seed, uint64_t philox_offset) {
+                                           uint64_t* seed_arr, uint64_t seed_val,
+                                           uint64_t* offset_arr, uint64_t offset_val) {
   const uint32_t bx = blockIdx.x, tx = threadIdx.x;
+
+  // Resolve seed/offset from tensor or scalar
+  uint64_t philox_seed = seed_arr ? seed_arr[0] : seed_val;
+  uint64_t philox_offset = offset_arr ? offset_arr[0] : offset_val;
```
Comment (yzh119) [inline include/flashinfer/sampling.cuh:1057]: ditto

File: include/flashinfer/sampling.cuh [hunk 5]
```diff
@@ -1098,10 +1128,16 @@ template <uint32_t BLOCK_THREADS, BlockScanAlgorithm SCAN_ALGORITHM,
...
-                                               uint64_t philox_offset) {
+                                               float top_p_val, uint32_t d, uint64_t* seed_arr,
+                                               uint64_t seed_val, uint64_t* offset_arr,
+                                               uint64_t offset_val) {
   const uint32_t batch_size = gridDim.x;
   const uint32_t bx = blockIdx.x, tx = threadIdx.x;
+
+  // Resolve seed/offset from tensor or scalar
+  uint64_t philox_seed = seed_arr ? seed_arr[0] : seed_val;
+  uint64_t philox_offset = offset_arr ? offset_arr[0] : offset_val;
```
Comment (yzh119) [inline include/flashinfer/sampling.cuh:1145]: ditto

File: include/flashinfer/sampling.cuh [hunk 6]
```diff
@@ -1744,9 +1783,15 @@ __global__ void ChainSpeculativeSampling(DType* draft_probs, IdType* draft_token
...
                                          uint32_t num_speculative_tokens, uint32_t d,
-                                         uint64_t philox_seed, uint64_t philox_offset) {
+                                         uint64_t* seed_arr, uint64_t seed_val,
+                                         uint64_t* offset_arr, uint64_t offset_val) {
   const uint32_t bx = blockIdx.x, tx = threadIdx.x;
   const uint32_t row_idx = bx;
+
+  // Resolve seed/offset from tensor or scalar
+  uint64_t philox_seed = seed_arr ? seed_arr[0] : seed_val;
+  uint64_t philox_offset = offset_arr ? offset_arr[0] : offset_val;
```
Comment (yzh119) [inline include/flashinfer/sampling.cuh:1799]: ditto

File: include/flashinfer/sampling.cuh [hunk 7]
```diff
@@ -742,9 +748,15 @@ template <uint32_t BLOCK_THREADS, BlockScanAlgorithm SCAN_ALGORITHM,
...
 __global__ void SamplingFromProbKernel(DType* probs, IdType* output, IdType* indices, uint32_t d,
-                                       uint64_t philox_seed, uint64_t philox_offset) {
+                                       uint64_t* seed_arr, uint64_t seed_val, uint64_t* offset_arr,
+                                       uint64_t offset_val) {
   curandStatePhilox4_32_10_t state;
   const uint32_t bx = blockIdx.x, tx = threadIdx.x;
+
+  // Resolve seed/offset from tensor or scalar
+  uint64_t philox_seed = seed_arr ? seed_arr[0] : seed_val;
+  uint64_t philox_offset = offset_arr ? offset_arr[0] : offset_val;
```
Comment (yzh119) [inline include/flashinfer/sampling.cuh:764]: ditto

File: include/flashinfer/sampling.cuh (hunk 1 above)
Comment (IzzyPutterman) [reply include/flashinfer/sampling.cuh:716]: Hmmm, I guess this PR only allows for CG compliance with passing in the seed and offset.  I would argue that its the user's responsibility to update the seed and offset as they use this function.  Otherwise there could be a lack of visibility into how the RNG is actually seeded.
No idea if this is a reasonable take, but I'm curious on your thoughts (and if I missed something super obvious lol).

Comment (IzzyPutterman) [reply include/flashinfer/sampling.cuh:716]: Spoke offline, confirmed this behavior is ok with the doc strings added.
